### PR TITLE
Adding a list implementation of applicative functors.

### PR DESCRIPTION
**Pull request by jfarid27**
PR for #2288

### CODE REVIEW

**Comment by akre54** [comment]
Can you give a more real-world example for this?

File: underscore.js [hunk 1]
```diff
@@ -919,6 +919,31 @@
...
+  // Returns non-deterministic map computation of list of functions applied to combinations of
+  // arguments from given lists of possible arguments. An implementation of lists as applicative functors.
+  _.nMap = function(funcs) {
+    var self = this;
+    var args = Array.prototype.slice.call(arguments, 1);
+
+    if (args.length === 1) {
+      return _.map(funcs, function(func) {
+        return _.map(args[0], func);
+      }).reduce(function(agg, result) {
```
Comment (akre54) [inline underscore.js]: Keep in mind the return value for `_.map` is an array. This `reduce` call here relies on `Array.prototype.reduce` which isn't present in pre-es5 environs. Try using `_.chain`.

File: underscore.js [hunk 2]
```diff
@@ -919,6 +919,31 @@
...
+        return agg.concat(result);
+      }, []);
+    } else if (args.length > 1) {
+      var partials = _.map(funcs, function(func) {
+        return _.map(args[0], function(arg){
+          return _.partial(func, arg);
+        });
+      }).reduce(function(agg, result) {
```
Comment (akre54) [inline underscore.js]: ditto

**Comment by jfarid27** [comment]
@akre54 I just added using _.reduce. Check out my write-up in #2288. The real world example is in the jsFiddle mentioned, but again, it's just a nice way to implement list comprehensions. Also, since a list comprehension can only "compute" one function, this is a beefed up list comprehension.

**Comment by akre54** [comment]
I saw #2288. Got something more-real world than adding and multiplying numbers? List comprehensions are great, but this looks really complicated.

Philosophically, Underscore tries to be a set of generally-applicable composable utilities and I'm struggling to see how this fits.

**Comment by jfarid27** [comment]
@akre54 That's a hard request. It's abstract, like map, reduce, partial application, etc. Specifically it's a way to do this in javascript, without ES7 support...

```python
x = [f(x, y) for f in fs for x in xs for y in ys]
```

In fact, map and this nMap are related. They look like this.
```haskell
map :: (a -> b) -> [a] -> [b]
nMap :: [(a -> b)] -> [a] -> [b]
```

It just allows for a list of functions to apply to a given list of a's.

Imagine asking what's a real world application of map? It just helps in certain situations. Map can be implemented with for loops, but I hate rolling my own when I need to do things like get combinations of values out of hash tables, or make lots of combinations of values to test functions on. [For me, adding and multiplying numbers is real world since I work on data, mainly doing sums on pivots of a hash table.](http://jsfiddle.net/e2jrbtor/1/) Which real world application would you like?

**Comment by akre54** [comment]
> For me, adding and multiplying numbers is real world since I work on data, mainly doing sums on pivots of a hash table.

Totally. But Underscore tries to give you the building blocks for function composition, not become the end-all be-all of functional libraries. It doesn't have a ton of other data-manipulation or statistics utilities, and adding this function without adding others wouldn't make a ton of sense without a few broader applicable use cases (which `map` and `partial` have).

You can do these examples a bit simpler (and IMO clearer) with nested `_.map` and `_.flatten` (see also the discussion on `flatMap` and `concatMap` on #452).

**Comment by jfarid27** [comment]
@akre54 I wish I saw #452 before sending this PR. Had I known you already had people clamoring at the gates to implement *every* functional pattern that Haskell and Scala, I probably wouldn't have worked on this...

**Pull request closed without merging**